File: script/make/protocol/js/WelfareProtocol.js
export default class WelfareProtocol {
    static encode(textEncoder, view, offset, protocol, data) {
        switch (protocol) {
            case 15001: {
                return new DataView(view.buffer.slice(0, offset));
            }
            case 15002: {
                // extend
                while (view.byteLength < offset + 2) {
                    const extendView = new DataView(new ArrayBuffer(view.byteLength * 2));
                    (new Uint8Array(extendView.buffer)).set(new Uint8Array(view.buffer));
                    view = extendView;
                }
                // 兑换码
                const dataArray = textEncoder.encode(data);
                view.setUint16(offset, dataArray.length, false);
                offset = offset + 2;
                // extend
                while (view.byteLength < offset + dataArray.length) {
                    const extendView = new DataView(new ArrayBuffer(view.byteLength * 2));
                    (new Uint8Array(extendView.buffer)).set(new Uint8Array(view.buffer));
                    view = extendView;
                }
                (new Uint8Array(view.buffer, offset)).set(dataArray);
                offset = offset + dataArray.length;
                return new DataView(view.buffer.slice(0, offset));
            }
            case 15003: {
                // extend
                while (view.byteLength < offset + 8) {
                    const extendView = new DataView(new ArrayBuffer(view.byteLength * 2));
                    (new Uint8Array(extendView.buffer)).set(new Uint8Array(view.buffer));
                    view = extendView;
                }
                // 红包编号
                view.setBigUint64(offset, data, false);
                offset = offset + 8;
                return new DataView(view.buffer.slice(0, offset));
            }
            case 15004: {
                // extend
                while (view.byteLength < offset + 8) {
                    const extendView = new DataView(new ArrayBuffer(view.byteLength * 2));
                    (new Uint8Array(extendView.buffer)).set(new Uint8Array(view.buffer));
                    view = extendView;
                }
                // 红包编号
                view.setBigUint64(offset, data, false);
                offset = offset + 8;
                return new DataView(view.buffer.slice(0, offset));
            }
            default: throw("unknown protocol define: " + protocol)
        }
    }

    static decode(textDecoder, view, offset, protocol) {
        switch (protocol) {
            case 15001: {
                // 结果
                const dataLength = view.getUint16(offset, false);
                offset = offset + 2;
                const dataArray = new Uint8Array(view.buffer.slice(offset, offset + dataLength));
                const data = textDecoder.decode(dataArray);
                offset = offset + dataLength;
                return data;
            }
            case 15002: {
                // 结果
                const dataLength = view.getUint16(offset, false);
                offset = offset + 2;
                const dataArray = new Uint8Array(view.buffer.slice(offset, offset + dataLength));
                const data = textDecoder.decode(dataArray);
                offset = offset + dataLength;
                return data;
            }
            case 15003: {
                // 
                // 红包编号
                const luckyMoneyNo = view.getBigUint64(offset, false);
                offset = offset + 8;
                // 总金币
                const totalGold = view.getBigUint64(offset, false);
                offset = offset + 8;
                // 总数量
                const totalNumber = view.getUint32(offset, false);
                offset = offset + 4;
                // 已经领取人数
                const receiveNumber = view.getUint16(offset, false);
                offset = offset + 2;
                // 领取列表
                const receiveList = [];
                let receiveListLength = view.getUint16(offset, false);
                offset = offset + 2;
                while (--receiveListLength >= 0) {
                    // 
                    // 服务器Id
                    const receiveListServerId = view.getUint16(offset, false);
                    offset = offset + 2;
                    // 角色Id
                    const receiveListRoleId = view.getBigUint64(offset, false);
                    offset = offset + 8;
                    // 角色名
                    const receiveListRoleNameLength = view.getUint16(offset, false);
                    offset = offset + 2;
                    const receiveListRoleNameArray = new Uint8Array(view.buffer.slice(offset, offset + receiveListRoleNameLength));
                    const receiveListRoleName = textDecoder.decode(receiveListRoleNameArray);
                    offset = offset + receiveListRoleNameLength;
                    // 金币
                    const receiveListGold = view.getBigUint64(offset, false);
                    offset = offset + 8;
                    // 领取时间
                    const receiveListTime = view.getUint32(offset, false);
                    offset = offset + 4;
                    // object
                    const receiveListLuckyMoneyRole = {"serverId": receiveListServerId, "roleId": receiveListRoleId, "roleName": receiveListRoleName, "gold": receiveListGold, "time": receiveListTime};
                    // add
                    receiveList.push(receiveListLuckyMoneyRole);
                }
                // 发送时间
                const time = view.getUint32(offset, false);
                offset = offset + 4;
                // object
                const luckyMoney = {"luckyMoneyNo": luckyMoneyNo, "totalGold": totalGold, "totalNumber": totalNumber, "receiveNumber": receiveNumber, "receiveList": receiveList, "time": time};
                return luckyMoney;
            }
            case 15004: {
                // 
                // 结果
                const resultLength = view.getUint16(offset, false);
                offset = offset + 2;
                const resultArray = new Uint8Array(view.buffer.slice(offset, offset + resultLength));
                const result = textDecoder.decode(resultArray);
                offset = offset + resultLength;
                // 金币
                const gold = view.getBigUint64(offset, false);
                offset = offset + 8;
                // object
                const data = {"result": result, "gold": gold};
                return data;
            }
            case 15005: {
                // 

                // object
                const data = {};
                return data;
            }
            default: throw("unknown protocol define: " + protocol)
        }
    }
}

File: script/make/protocol/cs/WelfareProtocol.cs
public static class WelfareProtocol
{
    public static void Encode(System.Text.Encoding encoding, System.IO.BinaryWriter writer, System.UInt16 protocol, System.Object data) 
    {
        switch (protocol) 
        {
            case 15001:
            {
                return;
            }
            case 15002:
            {
                // 兑换码
                var dataBytes = encoding.GetBytes((System.String)data);
                writer.Write(System.Net.IPAddress.HostToNetworkOrder((System.Int16)dataBytes.Length));
                writer.Write(dataBytes);
                return;
            }
            case 15003:
            {
                // 红包编号
                writer.Write(System.Net.IPAddress.HostToNetworkOrder((System.Int64)(System.UInt64)data));
                return;
            }
            case 15004:
            {
                // 红包编号
                writer.Write(System.Net.IPAddress.HostToNetworkOrder((System.Int64)(System.UInt64)data));
                return;
            }
            default:throw new System.ArgumentException(System.String.Format("unknown protocol define: {0}", protocol));
        }
    }

    public static System.Object Decode(System.Text.Encoding encoding, System.IO.BinaryReader reader, System.UInt16 protocol) 
    {
        switch (protocol) 
        {
            case 15001:
            {
                // 结果
                var dataLength = (System.UInt16)System.Net.IPAddress.NetworkToHostOrder(reader.ReadInt16());
                var data = encoding.GetString(reader.ReadBytes(dataLength));
                return data;
            }
            case 15002:
            {
                // 结果
                var dataLength = (System.UInt16)System.Net.IPAddress.NetworkToHostOrder(reader.ReadInt16());
                var data = encoding.GetString(reader.ReadBytes(dataLength));
                return data;
            }
            case 15003:
            {
                // 
                // 红包编号
                var luckyMoneyNo = (System.UInt64)System.Net.IPAddress.NetworkToHostOrder(reader.ReadInt64());
                // 总金币
                var totalGold = (System.UInt64)System.Net.IPAddress.NetworkToHostOrder(reader.ReadInt64());
                // 总数量
                var totalNumber = (System.UInt32)System.Net.IPAddress.NetworkToHostOrder(reader.ReadInt32());
                // 已经领取人数
                var receiveNumber = (System.UInt16)System.Net.IPAddress.NetworkToHostOrder(reader.ReadInt16());
                // 领取列表
                var receiveListLength = (System.UInt16)System.Net.IPAddress.NetworkToHostOrder(reader.ReadInt16());
                var receiveList = new System.Collections.Generic.List<System.Object>(receiveListLength);
                while (receiveListLength-- > 0)
                {
                    // 
                    // 服务器Id
                    var receiveListServerId = (System.UInt16)System.Net.IPAddress.NetworkToHostOrder(reader.ReadInt16());
                    // 角色Id
                    var receiveListRoleId = (System.UInt64)System.Net.IPAddress.NetworkToHostOrder(reader.ReadInt64());
                    // 角色名
                    var receiveListRoleNameLength = (System.UInt16)System.Net.IPAddress.NetworkToHostOrder(reader.ReadInt16());
                    var receiveListRoleName = encoding.GetString(reader.ReadBytes(receiveListRoleNameLength));
                    // 金币
                    var receiveListGold = (System.UInt64)System.Net.IPAddress.NetworkToHostOrder(reader.ReadInt64());
                    // 领取时间
                    var receiveListTime = (System.UInt32)System.Net.IPAddress.NetworkToHostOrder(reader.ReadInt32());
                    // object
                    var receiveListLuckyMoneyRole = new System.Collections.Generic.Dictionary<System.String, System.Object>() {{"serverId", receiveListServerId}, {"roleId", receiveListRoleId}, {"roleName", receiveListRoleName}, {"gold", receiveListGold}, {"time", receiveListTime}};
                    // add
                    receiveList.Add(receiveListLuckyMoneyRole);
                }
                // 发送时间
                var time = (System.UInt32)System.Net.IPAddress.NetworkToHostOrder(reader.ReadInt32());
                // object
                var luckyMoney = new System.Collections.Generic.Dictionary<System.String, System.Object>() {{"luckyMoneyNo", luckyMoneyNo}, {"totalGold", totalGold}, {"totalNumber", totalNumber}, {"receiveNumber", receiveNumber}, {"receiveList", receiveList}, {"time", time}};
                return luckyMoney;
            }
            case 15004:
            {
                // 
                // 结果
                var resultLength = (System.UInt16)System.Net.IPAddress.NetworkToHostOrder(reader.ReadInt16());
                var result = encoding.GetString(reader.ReadBytes(resultLength));
                // 金币
                var gold = (System.UInt64)System.Net.IPAddress.NetworkToHostOrder(reader.ReadInt64());
                // object
                var data = new System.Collections.Generic.Dictionary<System.String, System.Object>() {{"result", result}, {"gold", gold}};
                return data;
            }
            case 15005:
            {
                // 

                // object
                var data = new System.Collections.Generic.Dictionary<System.String, System.Object>() {};
                return data;
            }
            default:throw new System.ArgumentException(System.String.Format("unknown protocol define: {0}", protocol));
        }
    }
}